### PR TITLE
[ML] Add Detector config classes to protocol library

### PR DESCRIPTION
**Pull request by droberts195**
This commit adds the Detector class and its dependencies to the
X-Pack protocol library used by the high level REST client.

(Future commits will add the remaining config classes, plus results
and stats classes.)

These classes:

- Are immutable, with builders, but the builders do no validation
  beyond null checks
- Are convertible to and from X-Content, but NOT wire transportable
- Have lenient parsers to maximize compatibility across versions
- Have the same class names, member names and getter/setter names
  as the corresponding classes in X-Pack core to ease migration
  for transport client users
- Don't reproduce all the methods that do calculations or
  transformations that the the corresponding classes in X-Pack core
  have

### CODE REVIEW

**Comment by droberts195** [comment]
This is labelled `>non-issue` as it's useless on its own.  It's groundwork for future changes that will make ML endpoints available in the HLRC.

File: x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/job/config/Detector.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Defines the fields to be used in the analysis.
+ * <code>fieldname</code> must be set and only one of <code>byFieldName</code>
```
Comment (davidkyle) [inline x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/job/config/Detector.java]: Hmm I wonder how old the original comment is.

We should reference the docs here [detector config](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-job-resource.html#ml-detectorconfig) and [function reference](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-job-resource.html#ml-detectorconfig)

"Defines the fields and functions used in the analysis. A combination of <code>field_name</code>, <code>by_field_name</code> and <code>over_field_name</code> can be used depending on the specific function chosen. For more information see [configuring detectors](link) and [detector functions](link)"

**Review by davidkyle — COMMENTED** [review]
Looks good, I focused on the comments skimming the code assuming that it is a fairly straight forward port of the originals.

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**